### PR TITLE
Federated Posterior Averaging (FedPA)

### PR DESCRIPTION
**Pull request by alshedivat**
This PR adds the original implementation of FedPA in TFF.

I believe there's a similar internal CL (created in early Nov 2020) that wasn't merged into `google-research/federated` due to some build errors, which emerged because of fast-paced changes in TFF and this repository. In particular, FedPA heavily depends on the `optimization` subfolder which evolved over time. This PR has an up-to-date and fully working version compatible with the `HEAD` of this repository and TFF nightly.

### CODE REVIEW

**Review by rostamiz — APPROVED** [review]
Note: tests (which already exists internally) are to be modified added by rostami@ in follow-up CL.

File: posterior_averaging/main/federated_trainer.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+    Args:
+      model_fn: A no-arg function returning a `tff.learning.Model`.
+
+    Returns:
+      A `tff.templates.IterativeProcess`.
+    """
+    if FLAGS.task == 'shakespeare' or FLAGS.task == 'stackoverflow_nwp':
+
+      def client_weight_fn(local_outputs):
```
Comment (zcharles8) [inline posterior_averaging/main/federated_trainer.py:221]: I've seen little to no difference between 'num_tokens' and the default example weighting, so if you'd like to reduce complexity you could remove these two lines.

File: posterior_averaging/shared/fed_pa_schedule.py [hunk 2]
```diff
@@ -0,0 +1,820 @@
...
+  PostAvg to update weight deltas given a new approximate posterior sample.
+  Each field described below is a dictionary of the same structure as model's
+  trainable variables.
+
+  Fields:
+  - `num_samples`: The number of approximate posterior samples so far.
+  - `weights_delta`: Updates to model's trainable variables.
+  - `weights_sample_mean`: The mean of the all posterior samples so far.
+      `n`-th sample from the mean of the previous `(n - 1)` samples.
+  - `dp_state`: The state of the dynamic programming (DP) represented by a
```
Comment (zcharles8) [inline posterior_averaging/shared/fed_pa_schedule.py:101]: DP might be a bit confusing due to the prevalence of differential privacy. I don't think you necessarily need to change it, just wanted to call it out.

File: posterior_averaging/shared/fed_pa_schedule.py [hunk 3]
```diff
@@ -0,0 +1,820 @@
...
+  dp_state = attr.ib(default=None)
+
+  @classmethod
+  def from_weights(cls, initial_weights, updated_weights, num_samples=0):
+    """Creates a delta update output from intial and updated weights."""
+    # Initialize updates.
+    weights_delta = tf.nest.map_structure(lambda a, b: a - b, updated_weights,
+                                          initial_weights)
+    # Initial state for dynamic programming updates.
+    vk_tas_init = tf.nest.map_structure(
```
Comment (zcharles8) [inline posterior_averaging/shared/fed_pa_schedule.py:116]: Can we give this a clearer name? Generally I'm against "variable" names, like "v", or time-varying variable names, like "vk".

File: posterior_averaging/shared/fed_pa_schedule.py [hunk 4]
```diff
@@ -0,0 +1,820 @@
...
+      pass that contains the epoch loss, the final iterate of model's trainable
+      weights, and an approximate posterior sample of model's trainable weights.
+    previous_updates: A `DeltaUpdateOutputs` structure that contains information
+      about the previous updates to the model's trainable variables.
+    rho: A float that specifies the shrinkage coefficient.
+
+  Returns:
+    A `DeltaUpdateOutputs` structure.
+  """
+  n_1 = previous_updates.num_samples
```
Comment (zcharles8) [inline posterior_averaging/shared/fed_pa_schedule.py:290]: A more descriptive name would help here too, maybe "previous_num_examples"?

File: posterior_averaging/shared/fed_pa_schedule.py [hunk 5]
```diff
@@ -0,0 +1,820 @@
...
+                                un_flat)))
+
+  # Update DP state represented by TensorArrays.
+  i = tf.cast(n - 2, dtype=tf.int32)
+  vk_tas = tf.nest.map_structure(lambda vk_ta, vn: vk_ta.write(i, vn), vk_tas,
+                                 vn_flat)
+  dot_vk_uk_ta = dot_vk_uk_ta.write(i, dot_vn_un)
+
+  # Compute weights delta tilde: `weights_delta_tilde += coeff * vn / n`.
+  dot_wd_un = sum(
```
Comment (zcharles8) [inline posterior_averaging/shared/fed_pa_schedule.py:341]: Maybe we can separate this into multiple calls, instead of nesting?

**Review by zcharles8 — APPROVED** [review]
This looks really good. I left some minor comments about naming conventions, but they're generally optional. Let me know when you think everything is in good shape, and I'll create the internal PR.

File: posterior_averaging/shared/fed_pa_schedule.py (hunk 2 above)
Comment (alshedivat) [reply posterior_averaging/shared/fed_pa_schedule.py:101]: Good point! Renaming it to `recursion_state`, which should be more descriptive / less confusing.

File: posterior_averaging/shared/fed_pa_schedule.py (hunk 3 above)
Comment (alshedivat) [reply posterior_averaging/shared/fed_pa_schedule.py:116]: Ah, yes, I agree. The idea behind this notation was to match the equations that we provide in the paper, where `v_k` and `u_k` are particular intermediate vectors computed by our recursive algorithm. Let me think if I can improve the naming convention or maybe explain the naming convention beforehand in the docstring of the corresponding function.

**Comment by alshedivat** [comment]
Thanks for the review, @zcharles8! I pushed a few changes. I think everything should be in good shape now.

**Review by zcharles8 — APPROVED** [review]

**Pull request merged**